src/Build/Evaluation/Evaluator.cs

```diff
@@ -260,6 +266,10 @@ internal class Evaluator<P, I, M, D>
             _streamImports = new List<string>();
             // When the imports are concatenated with a semicolon, this automatically prepends a semicolon if and only if another element is later added.
             _streamImports.Add(string.Empty);
+
+            // Create a FileMatcher for the given combination of EvaluationContext and the project being evaluated.
+            IFileSystem fileSystem = _evaluationContext.GetFileSystemForProject(project);
+            _fileMatcher = new FileMatcher(fileSystem, evaluationContext.FileEntryExpansionCache);
```

Comment:
Why does the file matcher have to be sensitive to the Project as opposed to being encapsulated / hidden in the evaluation context?

Reply:
My understanding is that CPS maintains its cache per configured project so its content can be different depending on which project is being evaluated, even if they share one evaluation context. I'll let @lifengl / @arkalyanms confirm this requirement.

I have just pushed an updated based on yesterday's sync (too bad we did not record it) where I was asked to couple `IDirectoryCacheFactory` with the `Project` rather than with `EvaluationContext`. I do agree that having the cache be part of `EvaluationContext` would be cleaner but unfortunately it wouldn't work well for the intended use here.

Reply:
An alternative is for CPS to maintain multiple evaluation contexts, one for each cache "universe". I'm assuming that they don't have a unique cache per project (what's the point of having a unique cache per project?), but instead a few larger buckets, maybe based on different FS states?

Reply:
It is primary not for that reason, but couple issues we discussed.

1, we want to use this to capture project dependencies, so when things are changed, we can trigger reevaluation. For example, we can capture the project evaluation tries to load Directory.Build.Props file in the folder, although it does not exist, so when the user adds the file, we can trigger a new evaluation. Without something connecting to a specific project evaluation, we cannot do that.

2, for each project, we want to use IDirectoryCache beyond one evaluation context, because ReevaluateIfNecessary() can be called in any random place in CPS extensions. If DirectoryCache need be passed in each of them (as evaluationContext), so if it is called outside of CPS, it may not pass in evaluationContext, and it will hit disk on those calls, then others using the cache state.  This can lead into lots of problems because the cache might be slightly behind the real disk, and we can unpredictable behavior in the product scenarios. 


However, i noticed a potential problem later in the FindPredicate contract. Because the file/folder patterns are not passed to the CPS side, but it always scan all files/folders under the folder, it might cause excessive invalidation of the evaluation result. I am not sure whether it would become a real problem or not, mostly depending how SDK uses those glob patterns.

Reply:
Great point about the missing file/folder pattern. It's intentionally designed like this (i.e. that filtering is done MSBuild-side) so the logic stays in one place and we don't risk running into slightly different behavior depending on how it's implemented. Also, I don't believe that all patterns supported by MSBuild are handled by `System.IO` or the underlying filesystem calls so there would still be cases where invalidation would not be precise.

Can you use the result of `FindPredicate` to mark files/folders that are actually used by the project and should cause invalidation? In other words, items for which the predicate returns false would still be cached but if the file watcher fires for them the project will not be re-evaluated.